### ISSUE TITLE
Rich presence

### ISSUE THREAD

**Issue by Leeo97one** [flathub/com.discordapp.Discord#29]
How can we let Discord be notified by some rich prensence-compatible software? I guess it is more than just adding `--filesystem=/proc`.

**Comment by mjarkk** [comment]
## Same problem here  
  
After some digging in things that use discord's rich prensence i found they use a Websocket or a sock file.  
Most programs use the sock file located at `/run/user/1000/discord-ipc-0` (at least when not ran in a container)  

### Things i have tried to make it work:
- `--filesystem=/proc --filesystem=/run` and `--filesystem=/proc --filesystem=/run/user/1000` it will give a lot of errors probably because of incompatteble packages and other related things  
- `--filesystem=host`, `--filesystem=/proc --filesystem=/run/user/1000/discord-ipc-0` and `--filesystem=/proc --filesystem=/run/user/1000/discord-ipc-*` that also doesn't seem to work. discord rich prensence tells me it fails to connect probably because there is no file/dir created named `discord-ipc-0`
- `--share=ipc`, `--share=network` and `--socket=system-bus` will give errors, take long to start discord or just do nothing special  
  
### Some things i have found that maybe cause this to not work...  
- That number at the end (`0`) can be different according to the [nodejs Discord.js RPC Extension](https://github.com/discordjs/RPC/blob/master/src/transports/IPC.js#L16)  
- In that same file i see a view request to http://127.0.0.1 on port 6463 to 6472 (this is also dune in the Websocket connection file) discord's webserver might be also sandboxed with the result that you can't connect to discord  
- the `discord-ipc-0` file might be not in `/run/user/1000/` because that file gets placed in one of the following folders: `$XDG_RUNTIME_DIR`, `$TMPDIR`, `$TMP`, `$TEMP` or `/tmp`

### If someone want's to dive deeper than me here are maybe some usefull docs: 
- [flatpak sandbox permissions](http://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access)
- [flatpak sandbox permission reference](http://docs.flatpak.org/en/latest/sandbox-permissions-reference.html)
- [nodejs Discord.js RPC Extension](https://github.com/discordjs/RPC)

### My system details:
- OS: Fedora 
- `uname -r`: 4.17.3-200.fc28.x86_64
- Display server: Wayland

**Comment by TingPing** [comment]
- Exposing `/proc` at all is just broken conceptually and shouldn't be done.
- Exposing `xdg-run/discord-ipc-*` is not possible without exposing all of `xdg-run` which his just broken conceptually.

I don't see a reasonable solution to this sadly.

**Comment by mjarkk** [comment]
@gasinvein  
I used https://github.com/iCrawl/discord-vscode to test it because the code is fully open source and you can see what it's doing.  
That VScode extensions is build on top of: https://github.com/iCrawl/rpc if you want to dive deeper inside the code (This gives you probably more inside to what is happening than the extension)  

Here is a list of other things that use or have something to do with discord's Rich presence but i have no experience with these packages or how they work..
- [Arom Discord](https://atom.io/packages/atom-discord)
- [Sublime Discord RP](https://github.com/Snazzah/SublimeDiscordRP)
- [osu](https://osu.ppy.sh/home)
- [Virtual Trucker Rich Presence on Discord](https://github.com/VirtualTruckerRPC/Virtual-Trucker-Rich-Presence)
- [call of duty ww2](https://www.callofduty.com/nl/wwii)
- [Documentation](https://discordapp.com/developers/docs/rich-presence/how-to)

**Comment by gasinvein** [comment]
@mjarkk Well, I've managed to get it working with this tricks:
1. Bundle `socat` into Discord flatpak
2. Run discord with `--filesystem=xdg-run/discord:create`
3. Inside sandbox, run `socat UNIX-LISTEN:/run/user/1000/discord/ipc-0 UNIX-CONNECT:/run/user/1000/discord-ipc-0`
4. In the VS Code flatpak, create a symlink `/run/user/1000/discord-ipc-0` → `discord/ipc-0`, then run VS Code

**Comment by mjarkk** [comment]
@gasinvein amazing!!  
I'm definitely gonna use that trick.  
If no-one makes a flatpack package with this this included i'll probably make one myself with your fix

**Comment by gasinvein** [comment]
While it can be automated via wrappers, It's very hackish, especially because it requires modifications to all rpc-using apps, not just Discord.
Check out [my fork](/gasinvein/com.discordapp.Discord/tree/feature/rpc) and build it if you wish.

@TingPing your opinion?

**Comment by TingPing** [comment]
@gasinvein 	Placing it in `xdg-run/discord` sounds acceptable to me. It would be best to have a discussion with upstream Discord about this in general though. Of course I see no reason for them to actually care but maybe we get lucky.

**Comment by gasinvein** [comment]
@TingPing  If it's acceptable to start `socat` in background from Discord's wrapper, I'd open a PR.

**Comment by TingPing** [comment]
> Even if they'd care, changing socket path means breaking existing clients, so they definitely wouldn't do this.

Obviously they could keep both around for compat, deprecate the old one, and slowly move apps over.

> @TingPing If it's acceptable to start socat in background from Discord's wrapper, I'd open a PR.

As long as you make sure it dies with discord.

**Comment by gasinvein** [comment]
Maybe a readme would be better?
For host apps it's enough to do `ln -s discord/ipc-0 $XDG_RUNTIME_DIR/discord-ipc-0`.
And for other flatpacked apps it's required to also add `--filesystem=xdg-run/discord:create`. I think it's better to create the symlink via wrappers, so apps maintainers should be asked for this.

**Comment by TingPing** [comment]
We can make the README point to the wiki. I just think thats more convenient.

**Comment by gasinvein** [comment]
@TingPing [done](../wiki/Rich-Precense-(discord-rpc)), sorry for my bad english.

**Comment by phaux** [comment]
It doesn't work for me. There's nothing discord-related in `$XDG_RUNTIME_DIR` (`/run/user/1000`)

**Comment by lauracarlotta** [comment]
Boa noite, 

eu estou tentanfo entrar mas eu não estou conseguindo. Funciona pra Linux tamabém?

**Comment by D1360-64RC14** [comment]
I've found a file called `discord-ipc-0` inside `/run/user/1000/snap.discord/`. It seems like snap place these files inside an `snap.<app_name>` folder. There's a way to change the path where the extension will search?

**Comment by D1360-64RC14** [comment]
![image](https://user-images.githubusercontent.com/33406930/168307263-e1f8119a-31c7-43ba-aeba-f563993a4baf.png)
![image](https://user-images.githubusercontent.com/33406930/168307435-ab835577-ec42-473d-ae4b-d804c347a1f9.png)

Yeah, it worked with a symbolic link. Maybe adding a configuration to specify this path? Or a `vscode-snap: true` field?

**Comment by KHALILK23** [comment]
DISCORD RPC IS NOT WORKING

**Comment by Hacksore** [comment]
FWIW in https://github.com/flathub/flathub/pull/5450 I'm able to connect to RPC over websocket without issue.

I suspect that flatpak doesn't have as strict implications on the `127.0.0.1:[6463-6472]` websocket compared to the [unix socket documented here](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)).

**Comment by lionirdeadman** [comment]
> FWIW in [flathub/flathub#5450](https://github.com/flathub/flathub/pull/5450) I'm able to connect to RPC over websocket without issue.
> 
> I suspect that flatpak doesn't have as strict implications on the `127.0.0.1:[6463-6472]` websocket compared to the [unix socket documented here](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)).

Right, the `share=network` (https://github.com/flathub/com.discordapp.Discord/blob/master/com.discordapp.Discord.json#L17) permission allows for full network access - both internet and local. It would be nice to separate the two but this is not the case.

**Comment by sizwinz** [comment]
### Solution: Enabling Discord Rich Presence in Flatpak VS Code

If you are running **VS Code as a Flatpak** (`com.visualstudio.code`) alongside **Discord (Flatpak or Native)**, Rich Presence extensions fail because the Flatpak sandbox restricts access to host IPC sockets.

Here is the step-by-step fix to bridge the connection:

#### 1. Grant Filesystem Permissions to VS Code

Allow the VS Code Flatpak container to access both standard host IPC paths and Discord's Flatpak app runtime directory:

```bash
flatpak override --user \
  --filesystem='xdg-run/discord-ipc-*' \
  --filesystem='xdg-run/app/com.discordapp.Discord:create' \
  com.visualstudio.code

```

*(Note: If using shells like `fish`, single quotes around wildcards are required so local globs aren't expanded prematurely).*

#### 2. Symlink Discord's IPC Socket (If Discord is also a Flatpak)

Discord Flatpak places its IPC socket at `/run/user/1000/app/com.discordapp.Discord/discord-ipc-0`. Create a symlink at the standard host path expected by RPC clients:

```bash
ln -sf /run/user/1000/app/com.discordapp.Discord/discord-ipc-0 /run/user/1000/discord-ipc-0

```

#### 3. Resolve Extension Conflicts

Ensure you only have **one** Discord RPC extension installed (e.g., `leonardssh.vscord` or `icrawl.discord-vscode`). Having multiple active extensions running simultaneously will cause socket connection collisions (`COULD_NOT_CONNECT`).

#### 4. Restart the Flatpak Sandbox

Modifying Flatpak overrides requires a full container process termination (`Developer: Reload Window` in VS Code is not enough):

```bash
flatpak kill com.visualstudio.code

```

Reopen VS Code, and Rich Presence will connect successfully.
